### PR TITLE
feat(moe): add CuTe DSL SM90 BF16 MoE backend

### PR DESCRIPTION
**Pull request by Aneureka**
## 📌 Description

Add a CuTe-DSL BF16/FP16 fused-MoE backend for SM90 (Hopper), exposed as `cute_dsl_fused_moe_bf16` / `CuteDslBf16MoEWrapper`. It provides an unquantized alternative to `cutlass_fused_moe` on Hopper using the contiguous-grouped execution model: `moe_sort` (index maps only) → GEMM1 (cp.async gather + grouped WGMMA + fused SwiGLU epilogue) → GEMM2 (grouped WGMMA + fused router-scaled scatter-reduce finalize).

Design doc: `docs/design_docs/cute_dsl_moe_sm90.md`.

Highlights:
- Kernels (`flashinfer/fused_moe/cute_dsl/hopper/`): persistent warp-specialized SM90 kernels. GEMM1 fuses the token permute into the A gather (the permute is never materialized) and gates up/gate projections in registers via a 32-column weight interleave. GEMM2 fuses the top-k combine into a `cp.reduce.async.bulk` scatter (fused mode) or scatters unscaled rows for a bitwise-reproducible fixed-order combine (`use_fused_finalize=False`).
- CUDA-graph safe and PDL-enabled: static grids sized from `num_tokens` only, device-side tile exit, no tensormap updates (expert index is the TMA L-coordinate), event-based aux-stream memset overlap, PDL across the three ops.
- Autotuning (`sm90_tuner.py`): a 6-field tactic (`tile_size, gemm1_tile_n, gemm2_tile_n, gemm2_tile_k, gemm2_cluster_shape_mn, gemm2_raster_along_m`) tuned through the FlashInfer AutoTuner with balanced profiling inputs; the shape heuristic competes as an explicit candidate, so tuned dispatch is never worse than the default. EP sharding (`num_local_experts` / `local_expert_offset`) is supported.
- Benchmarks: `cute_dsl_bf16_moe` routine in `benchmarks/flashinfer_benchmark.py` (+ `--autotune_cache` support for tune-once/replay measurement).

Next steps:
- Extend the support surface, such as quantization, more activation types, etc
- Optimize performance, especially for large intermediate_size.

## 🔍 Related Issues

[#3521](https://github.com/flashinfer-ai/flashinfer/issues/3521)

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

Added (all passing on H200, 78 tests):

- `tests/moe/test_cute_dsl_bf16_gather_grouped_gemm.py` — standalone GEMM1 kernel tests (tile shapes, 2-warpgroup tiles, per-rank geometries).
- `tests/moe/test_cute_dsl_bf16_grouped_gemm_finalize.py` — standalone GEMM2 kernel tests: fused mode (bf16/fp16, tile-k 32/64, cluster (1,2), M-major raster, poisoned padding rows, zero-scale routes) and deterministic mode (expanded-order scatter).
- `tests/moe/test_cute_dsl_bf16_moe.py` — end-to-end: bf16/fp16 vs fp32 reference, tactic overrides, auto-selection, EP shards (including all-routed-outside-shard), autotune profiling + cached-winner recall, process-local compile reuse, deterministic-mode bitwise check, PDL on/off equivalence, CUDA-graph capture/replay with fresh routing, tiny/empty batches, fail-fast input validation.
- `tests/trace/…` — trace template + regenerated example JSON.

### Kernel Performance

Measured on H200 against `cutlass_fused_moe` (`base` variant — unquantized BF16) on identical tensors and routing. Setup:

- Each backend is autotuned once per suite with the winners stored to an `--autotune_cache` file; all measured rounds replay the stored tactics with autotune disabled, so between-round variance is execution-only.
- Each per-round value is the median of 50 CUPTI timings after 10 dry runs, with a cold L2 before every timing. Each cell is the median of three alternating backend-pair rounds (AB/BA/AB), reported only when the between-round CV is at most 5% for both backends.
- Workloads: routed-expert geometries from the models' published configs, TP in {1, 4}, unexpanded token counts T in {1, 256, 1024, 4096, 16384}.

Cells are the speedup over `cutlass_fused_moe` (values > 1 mean CuTe-DSL is faster):

| model (h / I global / E / top_k) | tp -> I/rank | T=1 | 256 | 1024 | 4096 | 16384 |
|---|---|---|---|---|---|---|
| Qwen3-30B-A3B (2048/768/128/8) | 1 -> 768 | 1.38x | 1.02x | 1.02x | 1.27x | 1.31x |
| | 4 -> 192 | 1.82x | 1.24x | 1.34x | 1.75x | 1.90x |
| Qwen3-235B-A22B (4096/1536/128/8) | 1 -> 1536 | 1.10x | 1.00x | 0.88x | 1.08x | 1.14x |
| | 4 -> 384 | 1.48x | 1.05x | 1.09x | 1.35x | 1.38x |
| Qwen3-Next-80B-A3B (2048/512/512/10) | 1 -> 512 | 1.70x | 0.98x | 1.03x | 1.17x | 1.48x |
| | 4 -> 128 | 2.30x | 1.17x | 1.26x | 1.73x | 1.82x |
| GLM-4.5-Air (4096/1408/128/8) | 1 -> 1408 | 1.14x | 1.01x | 0.89x | 1.04x | 1.13x |
| | 4 -> 352 | 1.48x | 1.09x | 1.07x | 1.11x | 1.11x |
| Kimi-K2 (7168/2048/384/8) | 1 -> 2048 | 1.08x | 1.00x | 1.03x | 1.02x | 1.12x |
| | 4 -> 512 | 1.35x | 1.01x | 1.04x | 1.12x | 1.18x |
| DeepSeek-V3 (7168/2048/256/8) | 1 -> 2048 | 1.08x | 1.00x | 1.05x | 0.86x | 1.03x |
| | 4 -> 512 | 1.33x | 1.02x | 1.10x | 1.04x | 1.21x |
| Mixtral-8x7B (4096/14336/8/2) | 1 -> 14336 | 1.07x | 0.85x | 0.86x | 0.89x | 0.95x |
| | 4 -> 3584 | 1.18x | 0.90x | 1.07x | 1.07x | 1.05x |

All 70 cells pass the variance gate; their geo-mean speedup is **1.16x**, and CuTe-DSL is faster in 59. The per-model geo-mean ranges from **0.98x** (Mixtral-8x7B) to **1.41x** (Qwen3-Next-80B-A3B); the advantage grows with TP (smaller per-rank I) and is largest at T=1 decode (up to 2.30x). Between-round CVs: CuTe-DSL median 0.12% / p95 2.45% / max 4.23%; baseline median 0.14% / p95 1.13% / max 4.09%.

### vLLM End-to-end performance

```
vllm bench throughput \
    --model Qwen/Qwen3-30B-A3B \
    --dataset-name random \
    --moe-backend triton/flashinfer_cutedsl \
    --num-warmups 128 -tp 1/2/4/8
```

| TP | Triton req/s | CuTe-DSL req/s | speedup |
|---|---:|---:|---:|
| 1 | 40.77 | 44.61 | **1.094x** |
| 2 | 64.62 | 69.56 | **1.076x** |
| 4 | 82.24 | 89.12 | **1.084x** |
| 8 | 88.05 | 95.52 | **1.085x** |

## Reviewer Notes

- Shape constraints: `hidden % 64 == 0`, `2I % 64 == 0`, `I % 32 == 0`
  (design doc §1); violations fail fast with descriptive errors.

## Summary by CodeRabbit

* **New Features**
  * Added experimental SM90/Hopper CuTe-DSL support for unquantized BF16/FP16 fused MoE workloads.
  * Exposed a fused MoE function and reusable wrapper with autotuning, deterministic execution, CUDA Graph support, and expert-parallel handling.
  * Added support for related benchmarking and trace workflows.

* **Documentation**
  * Added design documentation covering the SM90 fused MoE implementation and limitations.
  * Updated benchmarking guidance and MoE backend support details.

* **Tests**
  * Added numerical and end-to-end coverage for fused operations, execution modes, autotuning, and edge cases.

### CODE REVIEW

**Comment by aleozlx** [comment]
## PR Review Screening

**CI verdict:** ✅ auto-run ok
**Review category:** `live` (rule fired: C1.2 contains a net-new public interface)
**Blocking checks:** none
**Release blocker:** no
**Early stop:** no

### Security
| Q | Answer | Evidence |
|---|--------|----------|
| S1 injection/supply-chain | no | — |
| S2 template overwritten | no | All template headers present |
| S3 template obligations | met | Checklist complete; tests + perf data supplied |
| S4 agent-directing text | no | — |

### Packaging
| Q | Answer | Evidence |
|---|--------|----------|
| C1.1 external dependency bump | no | No requirements/pyproject/submodule/action changes |
| C1.2 public API changes | yes — **net-new interface** | see below |
| C1.3 AOT/trace registration | no gap | No new `gen_*_module()`; `TraceTemplate` + `tests/trace/fi_trace_out/cute_dsl_fused_moe_bf16_h2048_e128_topk8.json` added |

**C1.2 detail** — `flashinfer/fused_moe/cute_dsl/sm90_fused_moe.py`, re-exported via `flashinfer/fused_moe/cute_dsl/__init__.py`, `flashinfer/fused_moe/__init__.py`, `flashinfer/__init__.py`:

- net-new: `@supported_compute_capability([90]) @flashinfer_api(trace=…) def cute_dsl_fused_moe_bf16(x, token_selected_experts, token_final_scales, w1_weight, w2_weight, num_experts: int, top_k: int, num_local_experts: Optional[int] = None, local_expert_offset: int = 0, use_fused_finalize: bool = True, moe_output: Optional[torch.Tensor] = None, enable_pdl: bool = True, *, intermediate_buffer=None, tile_size=None, gemm1_tile_n=None, gemm2_tile_n=None, gemm2_tile_k=None, gemm2_cluster_shape_mn=None, gemm2_raster_along_m=None) -> torch.Tensor`
- net-new: `class CuteDslBf16MoEWrapper.__init__(num_experts, top_k, hidden_size, intermediate_size, num_local_experts=None, local_expert_offset=0, tile_size=None, output_dtype=torch.bfloat16, enable_pdl=True, use_fused_finalize=True, gemm1_tile_n=None, gemm2_tile_n=None)` / `.run(x, token_selected_experts, token_final_scales, w1_weight, w2_weight, tactic=None, moe_output=None, intermediate_buffer=None) -> torch.Tensor`
- extension (internal): `CuteDslMoEInputsHelper.__init__(…, seed=515)` → `(…, seed=515, tse_idx=2)` in `flashinfer/fused_moe/cute_dsl/_inputs_helper.py`

Boundary case: the new surface is deliberately narrow (SM90 only, bf16/fp16 unquantized, SwiGLU only, `hidden % 64 == 0` / `2I % 64 == 0` / `I % 32 == 0`), which sits close to a disposable backend-specific op — a maintainer may reasonably downgrade routing to `human`.

### Presentation
| Q | Answer | Evidence |
|---|--------|----------|
| C2.1 perf claim backed | yes | H200, per-model shapes, TP/T sweep, coefficient-of-variation gate and a reproducible path; ratios are an acceptable form of measurement |

### Implementation
| Q | Answer | Evidence |
|---|--------|----------|
| C3.1 experimental track declared | no | No `experimental` label, no `@flashinfer_experimental_api`, nothing under `flashinfer/experimental/` |
| C3.2 shared/durable areas touched | yes | `flashinfer/jit/moe_utils.py`: `gen_moe_utils_module()` `supported_major_versions=[10]` → `[9, 10]` (shared MoE JIT module); kernels themselves are new self-contained files |
| C3.3 tests match behavior | yes | 3 new SM90-gated files under `tests/moe/`, collected by the `tests/` default scope of the h100 unit-test lane; numerical refcheck + CUDA-graph/EP/determinism cases |
| C3.4 release-blocker candidate | no | Feature addition, no hang/crash/regression claim |

### Experimental track
| Q | Answer | Evidence |
|---|--------|----------|
| C4.1 isolated from common areas | n-a | C3.1 = no |

**Notes for the maintainer**
- The trace template tags this path `status:experimental` (`flashinfer/trace/templates/moe.py`) while nothing declares the experimental track — worth reconciling before merge.
- `docs/design_docs/cute_dsl_moe_sm90.md` cites `test_sm90_moe_kernel_cache.py`, which is not in this PR.
- The doc check bot reports `cute_dsl_fused_moe_bf16` missing from `docs/api/*.rst`.

<sub>Generated by flashinfer-pr-screen · rubric: docs/code_review_guidance.md · not a code review · AI screening can make mistakes — a maintainer's judgment supersedes this report.</sub>

**Review by aleozlx — APPROVED** [review]

File: tests/trace/example.py [hunk 1]
```diff
@@ -990,6 +991,41 @@
         renormalize=True,
     )
 
+# ── SM90 CuTe-DSL fused MoE (Qwen3-30B-A3B: E=128, topk=8, h=2048, i=768) ───
+# Unquantized bf16, pre-routed. SM90-only and JIT-built, so wrapped in
+# suppress(): the trace JSON dumps before the kernel launches, so the
+# definition files appear even when the kernel can't run here.
+with contextlib.suppress(Exception):
```
[the remark this answers is not in the document]
Comment (Aneureka) [reply tests/trace/example.py:998]: This follows the existing convention.

**Comment by Aneureka** [comment]
@flashinfer-bot run

**Comment by Aneureka** [comment]
@flashinfer-bot run

**Comment by Aneureka** [comment]
@flashinfer-bot run
